File: supermarket-pricing/python/tests/product_builder.py
from __future__ import annotations

from supermarket_pricing import (
    BuyOneGetOneFree,
    Money,
    MultiBuy,
    PricingRule,
    Product,
    UnitPrice,
    WeightedPrice,
)


class ProductBuilder:
    def __init__(self) -> None:
        self._sku = "X"
        self._name = "Default Item"
        self._pricing_rule: PricingRule = UnitPrice(Money(100))

    def with_sku(self, sku: str) -> ProductBuilder:
        self._sku = sku
        return self

    def named(self, name: str) -> ProductBuilder:
        self._name = name
        return self

    def with_unit_price(self, cents: int) -> ProductBuilder:
        self._pricing_rule = UnitPrice(Money(cents))
        return self

    def with_multi_buy(self, group_size: int, group_cents: int, item_cents: int) -> ProductBuilder:
        self._pricing_rule = MultiBuy(group_size, Money(group_cents), Money(item_cents))
        return self

    def with_buy_one_get_one_free(self, item_cents: int) -> ProductBuilder:
        self._pricing_rule = BuyOneGetOneFree(Money(item_cents))
        return self

    def with_weighted_price(self, cents_per_kg: int) -> ProductBuilder:
        self._pricing_rule = WeightedPrice(cents_per_kg)
        return self

    def with_pricing_rule(self, rule: PricingRule) -> ProductBuilder:
        self._pricing_rule = rule
        return self

    def build(self) -> Product:
        return Product(self._sku, self._name, self._pricing_rule)

File: supermarket-pricing/typescript/tests/ProductBuilder.ts
import { Money } from '../src/Money.js';
import type { PricingRule } from '../src/PricingRule.js';
import { BuyOneGetOneFree, MultiBuy, UnitPrice, WeightedPrice } from '../src/PricingRule.js';
import { Product } from '../src/Product.js';

export class ProductBuilder {
  private _sku = 'X';
  private _name = 'Default Item';
  private _pricingRule: PricingRule = new UnitPrice(new Money(100));

  withSku(sku: string): this { this._sku = sku; return this; }
  named(name: string): this { this._name = name; return this; }
  withUnitPrice(cents: number): this { this._pricingRule = new UnitPrice(new Money(cents)); return this; }
  withMultiBuy(groupSize: number, groupCents: number, itemCents: number): this {
    this._pricingRule = new MultiBuy(groupSize, new Money(groupCents), new Money(itemCents));
    return this;
  }
  withBuyOneGetOneFree(itemCents: number): this {
    this._pricingRule = new BuyOneGetOneFree(new Money(itemCents));
    return this;
  }
  withWeightedPrice(centsPerKg: number): this {
    this._pricingRule = new WeightedPrice(centsPerKg);
    return this;
  }
  withPricingRule(rule: PricingRule): this { this._pricingRule = rule; return this; }

  build(): Product { return new Product(this._sku, this._name, this._pricingRule); }
}
